pkg/translator/prometheusremotewrite/histograms.go

```diff
@@ -0,0 +1,167 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheusremotewrite // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/prometheusremotewrite"
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/value"
+	"github.com/prometheus/prometheus/prompb"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+const defaultZeroThreshold = 1e-128
+
+func addSingleExponentialHistogramDataPoint(
+	metric string,
+	pt pmetric.ExponentialHistogramDataPoint,
+	resource pcommon.Resource,
+	settings Settings,
+	series map[string]*prompb.TimeSeries,
+) error {
+	labels := createAttributes(
+		resource,
+		pt.Attributes(),
+		settings.ExternalLabels,
+		model.MetricNameLabel, metric,
+	)
+
+	sig := timeSeriesSignature(
+		pmetric.MetricTypeExponentialHistogram.String(),
+		&labels,
+	)
+	ts, ok := series[sig]
+	if !ok {
+		ts = &prompb.TimeSeries{
+			Labels: labels,
+		}
+		series[sig] = ts
+	}
+
+	histogram, err := exponentialToNativeHistogram(pt)
+	if err != nil {
+		return err
+	}
+	ts.Histograms = append(ts.Histograms, histogram)
+
+	exemplars := getPromExemplars[pmetric.ExponentialHistogramDataPoint](pt)
+	ts.Exemplars = append(ts.Exemplars, exemplars...)
+
+	return nil
+}
+
+// exponentialToNativeHistogram  translates OTel Exponential Histogram data point
+// to Prometheus Native Histogram.
+func exponentialToNativeHistogram(p pmetric.ExponentialHistogramDataPoint) (prompb.Histogram, error) {
+	scale := p.Scale()
+	if scale < -4 || scale > 8 {
+		return prompb.Histogram{},
+			fmt.Errorf("cannot convert exponential to native histogram."+
+				" Scale must be <= 8 and >= -4, was %d", scale)
+		// TODO: downscale to 8 if scale > 8
+	}
+
+	pSpans, pDeltas := convertBucketsLayout(p.Positive())
+	nSpans, nDeltas := convertBucketsLayout(p.Negative())
+
+	h := prompb.Histogram{
+		Schema: scale,
+
+		ZeroCount: &prompb.Histogram_ZeroCountInt{ZeroCountInt: p.ZeroCount()},
+		// TODO use zero_threshold, if set, see
+		// https://github.com/open-telemetry/opentelemetry-proto/pull/441
+		ZeroThreshold: defaultZeroThreshold,
+
+		PositiveSpans:  pSpans,
+		PositiveDeltas: pDeltas,
+		NegativeSpans:  nSpans,
+		NegativeDeltas: nDeltas,
+
+		Timestamp: convertTimeStamp(p.Timestamp()),
+	}
+
+	if p.Flags().NoRecordedValue() {
+		h.Sum = math.Float64frombits(value.StaleNaN)
```

Comment:
Does this mean if there was a reset in the histogram, we send a stale sample until we have new observations?

Reply:
`NoRecordedValue` indicates [explicitly missing data](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L324-L327) which corresponds to staleness marker. 

`NoRecordedValue` is handled in the same way  in other places in otlp to prw conversion logic, see add summary, hist, sum data points methods in [pkg/translator/prometheusremotewrite/helper.go]( https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/translator/prometheusremotewrite/helper.go)